### PR TITLE
TST: ndimage: generalize gaussian_filter axes tests

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
https://github.com/scipy/scipy/pull/18261

#### What does this implement/fix?
Here's my take on generalizing the existing `gaussian_filter` `axes` tests.

#### Additional information
I restored the original test file in the first commit, added my generalizations of the tests in the second commit, and added back your (unmodified) `test_minmax_nonseparable_axes` test in the third. Therefore, to minimize the diff, please see just the middle commit f772bc221004827f784eed6111f2f84842d3e0d7.

### CODE REVIEW

File: scipy/ndimage/tests/test_filters.py [hunk 1]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
+                              (ndimage.uniform_filter, 1, 3.0),
+                              (ndimage.minimum_filter, 1, 3.0),
+                              (ndimage.maximum_filter, 1, 3.0)])
     @pytest.mark.parametrize(
         'axes',
         tuple(itertools.combinations(range(-3, 3), 1))
         + tuple(itertools.combinations(range(-3, 3), 2))
         + ((0, 1, 2),))
-    def test_gauss_axes(self, axes):
+    def test_filter_axes(self, filter_func, size0, size, axes):
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:705]: Changes to this one are probably pretty easy to follow. Now we just need to parameterize over valid combinations of function name, the value of the `size`/`sigma` argument for the axis that is _not_ supposed to be filtered, the value of the `size`/`sigma` argument for the axes that _are_ filtered.

File: scipy/ndimage/tests/test_filters.py [hunk 2]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
+                        mode=['reflect', 'nearest', 'constant'])
+    kwargs_other = dict(origin=(-1, 0, 1),
+                        mode=['reflect', 'nearest', 'constant'])
+    @pytest.mark.parametrize("filter_func, size0, size, kwargs",
+                             [(ndimage.gaussian_filter, 0, 2, kwargs_gauss),
+                              (ndimage.uniform_filter, 1, 3, kwargs_other),
+                              (ndimage.maximum_filter, 1, 3, kwargs_other),
+                              (ndimage.minimum_filter, 1, 3, kwargs_other)])
+    @pytest.mark.parametrize('axes', itertools.combinations(range(-3, 3), 2))
+    def test_filter_axes_kwargs(self, filter_func, size0, size, kwargs, axes):
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:732]: This is very similar to the last one, except now we have other keyword arguments. With this new structure, it wouldn't be too hard to parameterized over all the same `axes` cases as the previous test (also include individual axes and all three axes, not just all possible pairs of axes).

File: scipy/ndimage/tests/test_filters.py [hunk 3]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
         array = numpy.arange(6 * 8 * 12, dtype=numpy.float64).reshape(6, 8, 12)
-        sigma = (1.0, 0.5)
-        radius = (4, 2)
-        order = (0, 1)
-        mode = ('reflect', 'nearest')
-        axes = tuple(ax % array.ndim for ax in axes)
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:745]: Previously, `axes` passed into the function were always non-negative integers. Was that intended, or did we mean to pass combinations involving negative integers, too?

File: scipy/ndimage/tests/test_filters.py [hunk 4]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
-        kwargs = dict(sigma=sigma, radius=radius, mode=mode, order=order)
-        if len(tuple(set(axes))) != len(axes):
+
+        kwargs = {key: np.array(val) for key, val in kwargs.items()}
+        axes = np.array(axes)
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]: Converting these to arrays so that we can pull out the values of they keyword arguments that correspond with the selected `axes` via fancy indexing.

File: scipy/ndimage/tests/test_filters.py [hunk 5]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
+        n_axes = axes.size
+
+        # form kwargs that specify only the axes in `axes`
+        reduced_kwargs = {key:val[axes] for key, val in kwargs.items()}
+        if len(set(axes % array.ndim)) != len(axes):
             # parametrized cases with duplicate axes raise an error
             with pytest.raises(ValueError, match="axes must be unique"):
-                ndimage.gaussian_filter(array, axes=axes, **kwargs)
+                filter_func(array, [size]*n_axes, axes=axes, **reduced_kwargs)
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py:744]: Note that now `axes` may have negative integers. (`% array.ndim` is only in the `if` statement above.)

File: scipy/ndimage/tests/test_filters.py [hunk 6]
```diff
@@ -692,67 +693,75 @@ def test_gauss_memory_overlap(self):
...
-        sigma_3d[axes[1]] = sigma[1]
-        radius_3d = [0,] * array.ndim
-        radius_3d[axes[0]] = radius[0]
-        radius_3d[axes[1]] = radius[1]
-        mode_3d = [0,] * array.ndim
-        mode_3d[axes[0]] = mode[0]
-        mode_3d[axes[1]] = mode[1]
-        order_3d = [0,] * array.ndim
-        order_3d[axes[0]] = order[0]
-        order_3d[axes[1]] = order[1]
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]: We no longer need all this because instead of forming all of these down here, we extracted only the values we needed above.

**Review by mdhaber — COMMENTED** [review]
Some self-review .

File: scipy/ndimage/tests/test_filters.py [hunk 7]
```diff
@@ -692,91 +693,75 @@ def test_gauss_memory_overlap(self):
...
         assert_array_almost_equal(output1, input)
 
-    @pytest.mark.parametrize(
-        'function_name', ['gaussian_filter', 'uniform_filter',
-                          'minimum_filter', 'maximum_filter'])
+    @pytest.mark.parametrize(('filter_func', 'size0', 'size'),
+                             [(ndimage.gaussian_filter, 0, 1.0),
+                              (ndimage.uniform_filter, 1, 3.0),
+                              (ndimage.minimum_filter, 1, 3.0),
+                              (ndimage.maximum_filter, 1, 3.0)])
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]: I think these are supposed to be integers. (Apparently, this works with floats that are integral.)
```suggestion
                              (ndimage.uniform_filter, 1, 3),
                              (ndimage.minimum_filter, 1, 3),
                              (ndimage.maximum_filter, 1, 3)])
```

File: scipy/ndimage/tests/test_filters.py [hunk 8]
```diff
@@ -692,91 +693,75 @@ def test_gauss_memory_overlap(self):
...
-    @pytest.mark.parametrize(
-        'axes', tuple(itertools.combinations(range(-3, 3), 2))
-    )
-    def test_gauss_axes_kwargs(self, axes):
+    kwargs_gauss = dict(radius=[4, 2, 3], order=[0, 1, 2],
+                        mode=['reflect', 'nearest', 'constant'])
+    kwargs_other = dict(origin=(-1, 0, 1),
+                        mode=['reflect', 'nearest', 'constant'])
+    @pytest.mark.parametrize("filter_func, size0, size, kwargs",
+                             [(ndimage.gaussian_filter, 0, 2, kwargs_gauss),
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]: For consistency with above.
```suggestion
                             [(ndimage.gaussian_filter, 0, 1.0, kwargs_gauss),
```

File: scipy/ndimage/tests/test_filters.py [hunk 9]
```diff
@@ -785,26 +770,37 @@ def test_filter_invalid_axes(self, axes, filter_func, kwargs):
...
-    def test_gauss_axes_tuple_length_mismatch(self, tuple_ndim, mismatched):
+            ndimage.gaussian_filter(array, size, axes=axes)
+
+    @staticmethod
+    def _cases_axes_tuple_length_mismatch():
+        # Generate combinations of filter function, valid kwargs, and
+        # keyword-value pairs for which the value will become with mismatched
+        # (invalid) size
+        filter_func = ndimage.gaussian_filter
+        kwargs = {'radius': 3, 'mode': 'constant', 'sigma': 1.0, 'order': 0}
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]:
```suggestion
        kwargs = dict(radius=3, mode='constant', sigma=1.0, order=0)
```

File: scipy/ndimage/tests/test_filters.py [hunk 10]
```diff
@@ -785,26 +770,37 @@ def test_filter_invalid_axes(self, axes, filter_func, kwargs):
...
+        for key, val in kwargs.items():
+            yield filter_func, kwargs, key, val
+
+        filter_funcs = [ndimage.uniform_filter, ndimage.minimum_filter,
+                        ndimage.maximum_filter]
+        kwargs = {'size': 3, 'mode': 'constant', 'origin': 0}
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]:
```suggestion
        kwargs = dict(size=3, mode='constant', origin=0)
```

File: scipy/ndimage/tests/test_filters.py (hunk 1 above)
Comment (grlee77) [reply scipy/ndimage/tests/test_filters.py:705]: `size`/`size0` may be a little confusing for Gaussian but I think it is okay. `sigma` is related to the size of the filter even if it isn't directly an integer size...

File: scipy/ndimage/tests/test_filters.py [hunk 11]
```diff
@@ -785,26 +770,37 @@ def test_filter_invalid_axes(self, axes, filter_func, kwargs):
             error_class = ValueError
             match = "out of range"
         with pytest.raises(error_class, match=match):
-            filter_func(array, axes=axes, **kwargs)
-
-    @pytest.mark.parametrize('tuple_ndim', [1, 3])
-    @pytest.mark.parametrize('mismatched',
-                             ['radius', 'mode', 'sigma', 'order'])
-    def test_gauss_axes_tuple_length_mismatch(self, tuple_ndim, mismatched):
+            ndimage.gaussian_filter(array, size, axes=axes)
```
Comment (grlee77) [inline scipy/ndimage/tests/test_filters.py]: why was this line changed back from `filter_func`?
```suggestion
            filter_func(array, size, axes=axes)
```

File: scipy/ndimage/tests/test_filters.py [hunk 12]
```diff
@@ -785,26 +770,37 @@ def test_filter_invalid_axes(self, axes, filter_func, kwargs):
...
+
+    @staticmethod
+    def _cases_axes_tuple_length_mismatch():
```
Comment (grlee77) [inline scipy/ndimage/tests/test_filters.py:776]: This is a nice approach, thanks!

File: scipy/ndimage/tests/test_filters.py [hunk 13]
```diff
@@ -3,6 +3,7 @@
 import itertools
 import math
 import numpy
+import numpy as np
```
Comment (grlee77) [inline scipy/ndimage/tests/test_filters.py]: Should we either just stick with numpy or change numpy->np everywhere in this file?

**Review by grlee77 — COMMENTED** [review]
Thanks, this looks good to me aside from a couple things I commented on

File: scipy/ndimage/tests/test_filters.py (hunk 1 above)
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:705]: I thought about picking a word for size that is different from both, like `scale`? But yeah, I think that the comments can help here.

File: scipy/ndimage/tests/test_filters.py (hunk 11 above)
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py]: Oops I don't know.

File: scipy/ndimage/tests/test_filters.py [hunk 14]
```diff
@@ -692,91 +693,75 @@ def test_gauss_memory_overlap(self):
...
-        sigma = (1.0, 0.5)
-        radius = (4, 2)
-        order = (0, 1)
-        mode = ('reflect', 'nearest')
-        axes = tuple(ax % array.ndim for ax in axes)
-        kwargs = dict(sigma=sigma, radius=radius, mode=mode, order=order)
-        if len(tuple(set(axes))) != len(axes):
+
+        kwargs = {key: np.array(val) for key, val in kwargs.items()}
+        axes = np.array(axes)
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]:
```suggestion
        kwargs = {key: numpy.array(val) for key, val in kwargs.items()}
        axes = numpy.array(axes)
```

File: scipy/ndimage/tests/test_filters.py [hunk 15]
```diff
@@ -692,91 +693,75 @@ def test_gauss_memory_overlap(self):
...
-        order_3d[axes[0]] = order[0]
-        order_3d[axes[1]] = order[1]
-        kwargs = dict(sigma=sigma_3d, radius=radius_3d, mode=mode_3d,
-                      order=order_3d)
-        expected = ndimage.gaussian_filter(array, **kwargs)
+
+        output = filter_func(array, [size]*n_axes, axes=axes, **reduced_kwargs)
+
+        # result should be equivalent to sigma=0.0/size=1 on unfiltered axes
+        size_3d = np.full(array.ndim, fill_value=size0)
```
Comment (mdhaber) [inline scipy/ndimage/tests/test_filters.py]:
```suggestion
        size_3d = numpy.full(array.ndim, fill_value=size0)
```

File: scipy/ndimage/tests/test_filters.py (hunk 13 above)
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py]: To avoid a larger diff.
```suggestion
import numpy
```

File: scipy/ndimage/tests/test_filters.py (hunk 1 above)
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:705]:
```suggestion
    def test_filter_axes(self, filter_func, size0, size, axes):
        # Note: `size` is called `sigma` in `gaussian_filter`
```

**Comment by mdhaber** [comment]
Thanks @grlee77. I think I addressed all the comments. I ran the tests locally, and they still pass.

**Review by grlee77 — APPROVED** [review]
Thanks, looks good to me
